Optimize exhaustive_L2sqr_blas_cmax with AVX-512 intrinsics.

mulugetam opened the pull request:
Add a template specialization of `exhaustive_L2sqr_blas_cmax` for `SIMDLevel::AVX512` in `distances_avx512.cpp`, alongside the existing `AVX2` and `ARM_SVE` specializations. The new kernel doubles the SIMD width of the post-`sgemm` fixup pass from 8 to 16 lanes, processing 32 elements per inner-loop iteration instead of 16.

**Benchmark**

`benchs/bench_index_flat.py` delivers up to a **1.16x** speedup compared to the AVX2 implementation.

```
+---------+----------+------------+---------+
|    d    | AVX2 k=1 | AVX512 k=1 | Speedup |
+---------+----------+------------+---------+
|   128   | 0.181814 |   0.156076 |  1.16x  |
|   256   | 0.309386 |   0.288921 |  1.07x  |
|   768   | 0.836036 |   0.814894 |  1.03x  |
+---------+----------+------------+---------+
```

**K-means** on `SIFT1M`, using `benchs/bench_all_ivf/bench_kmeans.py`, achieves up to a **1.06x** speedup relative to the existing `AVX2` implementation.

mdouze commented:
Thanks for the contribution. We are in the process of overhauling the SIMD support for Faiss, so we may prefer to defer integrating this PR. CC: @subhadeepkaran

mulugetam committed:
Add AVX-512 specialization for exhaustive_L2sqr_blas_cmax

Add a template specialization of exhaustive_L2sqr_blas_cmax for
SIMDLevel::AVX512 in distances_avx512.cpp, alongside the existing
AVX2 and ARM_SVE specializations. The new kernel doubles the SIMD
width of the post-sgemm fixup pass from 8 to 16 lanes, processing
32 elements per inner-loop iteration instead of 16, and replaces
the AVX2 blendv pattern with the cleaner AVX-512 mask-blend idiom
(_mm512_cmp_ps_mask + _mm512_mask_blend_ps/_epi32).

Extends the dispatch site in distances.cpp to route SIMDLevel::AVX512
through the new specialization. Switches the SIMD level mask from
AVAILABLE_SIMD_LEVELS_A2 to AVAILABLE_SIMD_LEVELS_A1 so AVX512 is
included in the dispatcher's switch; AVX512_SPR falls through to
the AVX512 case via the existing dispatch fallthrough chain.

The optimization targets the k=1 BLAS-accelerated search path used
by IndexFlatL2.search and Clustering training. It runs when:
  - k == 1 (selects Top1BlockResultHandler)
  - no IDSelector is in use
  - nx * d >= distance_compute_blas_threshold (128000 by default)
  - d > 32 (otherwise exhaustive_L2sqr_fused_cmax preempts)
  - SIMDConfig::level is AVX512 or AVX512_SPR

Signed-off-by: Mulugeta Mammo <mulugeta.mammo@intel.com>

mulugetam commented:
@mnorris11 Could you please review this as well? Thanks!

mnorris11 commented on faiss/utils/simd_impl/distances_avx512.cpp:
```diff
@@ -1104,4 +1133,208 @@ int fvec_madd_and_argmin<SIMDLevel::AVX512>(
+                       ip_block.get(),
+                       &nyi);
+            }
+            for (int64_t i = i0; i < (int64_t)i1; i++) {
```
1. Should this inner loop be parallelized like it is in AVX2 and SVE versions? 
2. (nit) There are a few places where the AVX2 version uses static_cast and this version uses (int64_t).

mnorris11 replied:
On local benchmark, I see avx512 regress without pragma, probably good to add. Otherwise LGTM!

mulugetam commented:
@mnorris11 rebased with changes.

mnorris11 committed:
Merge branch 'main' into exhuastive_l2sqr
